
# Dataset Information:

- The dataset is acquired from `The Energy Institute`, you can load this data from the following link at https://www.energyinst.org/statistical-review/resources-and-data-downloads.

- The data offers details data on global energy production, consumption, and emissions, from various energy sources such as oil, natural gas, coal, nuclear, hydroelectric, and renewables over spans multiple years.


In [23]:
import mysql.connector
from mysql.connector import errorcode
import csv
import sqlalchemy as sq
# In the lines below and fill in any relevant details, such as if you set up a different user or password
# Importing keys
import config

PASSWORD=config.sql_key
myconnection = mysql.connector.connect(user='root', password=PASSWORD,
                                host='127.0.0.1', database='data604',port=3306)


In [24]:
import pandas as pd
import numpy as np
df = pd.read_csv('Ren_gen_energy_data.csv')

In [6]:
df.head()

Country   Wind_2022  Solar_2022  Hydro_2022  OtherRenewables_2022   
0     Canada   38.376387    7.050284  397.725241              9.123262  \
1     Mexico   20.528750   20.342040   35.558850              6.696620   
2         US  438.684200  207.150505  251.273544             71.509097   
3  Argentina   14.164993    2.940079   22.769590              2.352800   
4     Brazil   81.631569   30.126462  427.113549             55.033215   

    Wind_2023  Solar_2023  Hydro_2023  OtherRenewables_2023  
0   38.936293    7.641602  364.196497             10.384484  
1   21.650982   27.143349   20.399428              7.099758  
2  429.530709  240.525253  236.321296             67.290103  
3   14.475425    3.272582   29.898193              2.275717  
4   95.508935   51.482100  428.653200             55.832700

In [7]:
# Creating a cursor 
create_cursor = myconnection.cursor()

# CREATE TABLE
create_statement = '''
create table renewable_energy (
    Country varchar(50) NOT NULL,
    Wind_2022 float,
    Solar_2022 float,
    Hydro_2022 float,
    OtherRenewables_2022 float,
    Wind_2023 float,
    Solar_2023 float,
    Hydro_2023 float,
    OtherRenewables_2023 float
);
'''

try:
    create_cursor.execute(create_statement)
except mysql.connector.Error as err:
    if err.errno == errorcode.ER_TABLE_EXISTS_ERROR:
        print("Table already exists.")
    else:
        print(err.msg)
else:
    print("Table created successfully!")

create_cursor.close()

Table already exists.


True

In [8]:
# Column names
PATH = "Correctly_Aligned_Renewable_Energy_Data.csv"
with open(PATH, 'r') as file:
    reader = csv.reader(file)
    column_names = next(reader)

print(column_names)


['Country', 'Wind_2022', 'Solar_2022', 'Hydro_2022', 'OtherRenewables_2022', 'Wind_2023', 'Solar_2023', 'Hydro_2023', 'OtherRenewables_2023']


In [ ]:
# Inserting data into the table
insert_cursor = myconnection.cursor()

# Prepare column names and values
columnString = "`,`".join([str(col) for col in column_names])
# print(columnString)

with open(PATH, mode='r') as csv_file:
    csv_reader = csv.reader(csv_file)
    next(csv_reader)
    for row in csv_reader:
        # print(row)      
        insertCommand = f"INSERT INTO `renewable_energy` (`{columnString}`) VALUES (" + "%s," * (len(row) - 1) + "%s)"
        # print(insertCommand)
        insert_cursor.execute(insertCommand, tuple(row))
        
print("Data inserted successfully!")

# Commiting the changes
myconnection.commit()
insert_cursor.close()

Data inserted successfully!


# Queries

#### Understanding the dataset:

All numerical values in the dataset are represented in TWh (Terawatt-Hours).
A **Terawatt-Hour (TWh)** is a unit of energy equivalent to `one trillion watts` $(10^{12} watts)$ consumed or generated for one hour.
- One TWh could power:
    - The entire state of California for about 1.5 weeks.
    - 100 million homes for an hour.
    - The world’s data centers (which power the internet) for about 15 days.
- In 2021, Canada produced *625.7 terawatt-hours (TWh)* of electricity. 
- More than half of the electricity in Canada (60%) is generated from **hydroelectricity** in 20

In [11]:
# Engine
engine = sq.create_engine('mysql+mysqlconnector://root:Malharsql@127.0.0.1/data604', echo=False)

#### 1. Total Renewable Energy Production by Country
- This query calculates the total renewable energy production for each country in 2023 and 2022.
- The result is sorted in descending order by the total renewable energy production in 2023/2022.
- Highlights the countries leading in total renewable energy production.


In [ ]:
query = '''
SELECT Country, 
       (Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023) AS Total_Renewable_2023,
        (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022) AS Total_Renewable_2022
FROM renewable_energy
ORDER BY Total_Renewable_2023 DESC;
'''
# Ordering by 2023 totals in descending order to identify leaders


df_temp = pd.read_sql(query, engine)
print(df_temp.head())
del df_temp

  Country  Total_Renewable_2023  Total_Renewable_2022
0   China           2894.140732           2670.587265
1      US            973.667343            968.617340
2  Brazil            631.476933            593.904800
3  Canada            421.158884            452.275186
4   India            381.964584            378.705719


#### 2. Contribution of Renewable Energy Types in Canada (2023)
- Which renewable type (wind, solar, hydro, others) contributes to Canada’s total renewable energy in 2023.
- Highlights the Canada’s dominant renewable energy source in 2023.
- I will use this to compare the report which mentions that more than half of the electricity in Canada (60%) is generated from hydroelectricity in 2021.


In [13]:
# Query
query = '''
SELECT Country,
       Wind_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100 AS Wind_Contribution, 
       Solar_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Solar_Contribution,
       Hydro_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Hydro_Contribution, 
       OtherRenewables_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Other_Contribution
FROM renewable_energy
WHERE Country = 'Canada';
'''

df_temp = pd.read_sql(query, engine)
df_temp.head()

Country  Wind_Contribution  Solar_Contribution  Hydro_Contribution   
0  Canada           9.245037            1.814423           86.474848  \

   Other_Contribution  
0            2.465693

- From the above query, we can see that Canada's renewable energy mix in 2023 is 9.25% wind, 1.81% solar, 86.47% hydro, and 2.47% other renewables.
- We can also see that the contribution of hydro has increased from 2021 to 2023. This is because the hydro contribution in 2021 was 60% and in 2023 it is 86.47%.

#### 3. Percentage Change in Total Renewable Energy by Country
- Growth rate of total renewable energy production for each country from 2022 to 2023.
- We can identifies nations who are investing more in renewable energy production.

In [16]:
# Percentage Change in Total Renewable Energy by Country
query = '''
SELECT Country, 
       ((Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023) - 
        (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022)) * 100.0 /
        (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022) AS Percentage_Change
FROM renewable_energy
WHERE (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022) > 0
ORDER BY Percentage_Change DESC;
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(10)

Country  Percentage_Change
0                  Qatar         258.582080
1           Saudi Arabia         152.885988
2   United Arab Emirates          78.380510
3               Slovenia          56.495215
4                   Iran          55.127892
5        North Macedonia          36.193001
6                Croatia          33.611890
7             Uzbekistan          33.094284
8  Other Southern Africa          27.902717
9            Switzerland          26.960973

From the above query, we can see that the top three country with the highest percentage change in total renewable energy from 2022 to 2023 is:
- `Qatar`
- `Saudi Arabia`
- `United Arab Emirates`

What is interesting is that these countries are located in the Middle East which is rich in solar energy resources.
Let's see the contribution of renewable energy types in these countries in 2023.

### 4. Solar Energy Contribution in Qatar, UAE, and Saudi Arabia (2023)
- Contribution of renewable energy sources in countries like Qatar, UAE, and Saudi Arabia.
- Shows how these countries' renewable energy portfolios are shaped by their geographical location.

In [17]:
query = '''
SELECT Country,
       Wind_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100 AS Wind_Contribution, 
       Solar_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Solar_Contribution,
       Hydro_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Hydro_Contribution, 
       OtherRenewables_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Other_Contribution
FROM renewable_energy
WHERE Country IN ('Qatar', 'Saudi Arabia', 'United Arab Emirates')
ORDER BY Solar_Contribution DESC;
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(10)

# As expected, these countries have a high contribution of solar energy to the total renewable energy in 2023.

Country  Wind_Contribution  Solar_Contribution   
0  United Arab Emirates           0.389944           99.369958  \
1                 Qatar           0.000000           90.762137   
2          Saudi Arabia          25.133279           74.866721   

   Hydro_Contribution  Other_Contribution  
0                 0.0            0.240097  
1                 0.0            9.237863  
2                 0.0            0.000000

As expected, these countries have a high contribution of solar energy to the total renewable energy in 2023.

#### 5. Ranking Countries by Renewable Energy Types

- Ranks countries based on each type of renewable energy in 2023.
- Highlights contries dominating in specific renewable energy categories.

In [22]:

query = '''
SELECT Country, 
       Wind_2023, 
       RANK() OVER (ORDER BY Wind_2023 DESC) AS Wind_Rank,
       Solar_2023, 
       RANK() OVER (ORDER BY Solar_2023 DESC) AS Solar_Rank,
       Hydro_2023, 
       RANK() OVER (ORDER BY Hydro_2023 DESC) AS Hydro_Rank,
       OtherRenewables_2023, 
       RANK() OVER (ORDER BY OtherRenewables_2023 DESC) AS Other_Rank
FROM renewable_energy;
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(10)

Country  Wind_2023  Wind_Rank  Solar_2023  Solar_Rank  Hydro_2023   
0           China   885.8700          1   584.15000           1   1226.0000  \
1              US   429.5310          2   240.52500           2    236.3210   
2          Brazil    95.5089          4    51.48210           6    428.6530   
3         Germany   142.1030          3    61.21600           5     19.6390   
4           Japan    10.0090         27    96.99360           4     74.4993   
5       Indonesia     0.4810         60     0.71287          66     24.5900   
6           India    82.1064          5   113.40900           3    149.1670   
7  United Kingdom    81.9895          6    13.82650          17      5.1945   
8           Italy    23.5304         15    31.22740           9     38.9234   
9          Turkey    34.0697         11    20.51890          15     63.8542   

   Hydro_Rank  OtherRenewables_2023  Other_Rank  
0           1              198.1250           1  
1           4               67.2901           2  
2           2               55.8327           3  
3          37               49.4910           4  
4          11               42.0452           5  
5          33               39.4105           6  
6           6               37.2816           7  
7          51               33.9900           8  
8          19               20.3103           9  
9          14               19.4901          10

### REFERENCES:

The Sustainable Choice. (n.d.). Making sense of energy units, from https://thesustainablechoice.com/making-sense-of-energy-units/

Canada Energy Regulator. (n.d.). Provincial and territorial energy profiles – Canada, from https://www.cer-rec.gc.ca/en/data-analysis/energy-markets/provincial-territorial-energy-profiles/provincial-territorial-energy-profiles-canada.html

